libsignal-service/src/sender.rs

```diff
@@ -605,53 +617,35 @@ where
         recipient: &ServiceAddress,
         unidentified_access: Option<&SenderCertificate>,
         content: &[u8],
+        is_sync_message: bool,
     ) -> Result<Vec<OutgoingPushMessage>, MessageSenderError> {
         let mut messages = vec![];
-        let mut devices = vec![];
 
-        let myself = *recipient == self.local_address;
-        if !myself || unidentified_access.is_some() {
-            trace!("sending message to default device");
-            messages.push(
-                self.create_encrypted_message(
-                    recipient,
-                    unidentified_access,
-                    DEFAULT_DEVICE_ID.into(),
-                    content,
-                )
-                .await?,
-            );
-        } else {
-            devices.push(DEFAULT_DEVICE_ID);
-        }
+        let mut devices: HashSet<DeviceId> = self
+            .protocol_store
+            .get_sub_device_sessions(recipient)
+            .await?
+            .into_iter()
+            .map(DeviceId::from)
+            .collect();
 
-        devices.extend(
-            self.protocol_store
-                .get_sub_device_sessions(recipient)
-                .await?,
-        );
-        devices.sort_unstable();
-
-        let original_device_count = devices.len();
-        devices.dedup();
-        if devices.len() != original_device_count {
-            log::warn!("SessionStoreExt::get_sub_device_sessions should return unique device id's, and should not include DEFAULT_DEVICE_ID.");
-        }
+        // always send to the primary device no matter what
+        devices.insert(DEFAULT_DEVICE_ID.into());
 
-        // When sending to ourselves, don't include the local device
-        if myself {
-            devices.retain(|id| DeviceId::from(*id) != self.device_id);
+        // when sending a sync message, we always send it to all devices except the current one
+        if is_sync_message {
+            devices.remove(&self.device_id);
         }
 
         for device_id in devices {
-            trace!("sending message to device {}", device_id);
-            let ppa = get_preferred_protocol_address(
-                &self.protocol_store,
-                recipient,
-                device_id.into(),
-            )
-            .await?;
-            if self.protocol_store.load_session(&ppa).await?.is_some() {
+            debug!("sending message to device {}", device_id);
+            if device_id == DEFAULT_DEVICE_ID.into()
+                || self
+                    .protocol_store
+                    .load_session(&recipient.to_protocol_address(device_id))
+                    .await?
+                    .is_some()
```

Comment:
This is a bit outside the scope of the PR, but why do we have this check in place? `create_encrypted_message` already checks and creates session if needed. Is that behavior avoided here for some specific reason? It would also bring down the repeated session loading, saving a few CPU cycles.

Edit: Okay, this is about creating the message content and not actually sending it, that's why. But doesn't auto-creating a message make sense anyway? If there is no session, the message won't be sent to that device, right..?

Reply:
This is what we were doing before but I rewrote the code for a reason that I ignore 😄